### PR TITLE
Sanity check@0.1.0

### PR DESCRIPTION
**Pull request by edsu7**
See sanity-check.readme for tested scenarios

### CODE REVIEW

File: sanity-check/README.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
+# Package differentiate-json
+
+Updates provided `file_info_tsv` with additional fields from Clinical. This step is to avoid mismatching info for the same field between Clinical.SONG and Submission.SONG.
+
```
Comment (lindaxiang) [inline sanity-check/README.md:4]: `experiment_info_tsv`?

File: sanity-check/README.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+## Inputs
+
+`File_info_tsv` - See argo-data-submission-wf
```
Comment (lindaxiang) [inline sanity-check/README.md]: `experiment_info_tsv`

File: sanity-check/README.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
+# Package differentiate-json
```
Comment (lindaxiang) [inline sanity-check/README.md]: Package sanity-check?

File: sanity-check/README.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+## Usage
+
+### Run the package directly
+
+With inputs prepared, you should be able to run the package directly using the following command.
+Please replace the params file with a real one (with all required parameters and input files). Example
+params file(s) can be found in the `tests` folder.
+
+```
+nextflow run icgc-argo/argo-data-submission/differentiate-json/main.nf -r differentiate-json.v0.1.0 -params-file <your-params-json-file>
```
Comment (lindaxiang) [inline sanity-check/README.md]: differentiate-json.v0.1.0?

File: sanity-check/main.nf [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+params.cpus = 1
+params.mem = 1  // GB
+params.publish_dir = ""  // set to empty string will disable publishDir
+
+// payloadGenSeqExperiment
+params.song_url="https://submission-song.rdpc-qa.cancercollaboratory.org/"
+params.clinical_url="https://clinical.qa.argo.cancercollaboratory.org"
+params.api_token=""
+params.experiment_info_tsv="NO_FILE1"
+params.skip_sanity_check="NO_FILE2"
```
Comment (lindaxiang) [inline sanity-check/main.nf]: skip_sanity_check: use `boolean` instead of `NO_FILE`? Is`skip_dup_check` more proper?

File: sanity-check/main.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+    parser = argparse.ArgumentParser(description='Tool: sanity-check')
+    parser.add_argument('-t', '--api_token', dest='api_token', type=str,
+                        help='Platform UUID token', required=True)
+    parser.add_argument('-c', '--clinical_url', dest='clinical_url', type=str,
+                        help='Clinical SONG API interaction URL', required=True)
+    parser.add_argument('-s', '--submission_song_url', dest='submission_song_url',type=str,
+                        help='Submission SONG API interaction URL', required=True)
+    parser.add_argument('-x', '--experiment_info_tsv', dest='experiment_info_tsv', type=str,required=True,
+                        help='TSV contain experiment info and submitter IDs')
+    parser.add_argument('-f', '--force', action='store_true',help='JSON file containing metadata payload')
```
Comment (lindaxiang) [inline sanity-check/main.py]: `JSON file containing metadata payload`?

File: sanity-check/main.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    Python implementation of tool: sanity-check
+
+    This is auto-generated Python code, please update as needed!
+    """
+
+    parser = argparse.ArgumentParser(description='Tool: sanity-check')
+    parser.add_argument('-t', '--api_token', dest='api_token', type=str,
+                        help='Platform UUID token', required=True)
+    parser.add_argument('-c', '--clinical_url', dest='clinical_url', type=str,
+                        help='Clinical SONG API interaction URL', required=True)
```
Comment (lindaxiang) [inline sanity-check/main.py]: Clinical API interaction URL

File: sanity-check/main.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+            args.clinical_url,
+            args.api_token
+        )
+        
+        final_metadata=compile_metadata(
+            metadata,
+            clinical_metadata
+        )
+        if not args.force:
+            check_study_exists(
```
Comment (lindaxiang) [inline sanity-check/main.py]: `check_study_exists` should be always checked.

File: sanity-check/main.py [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+        sys.exit("Unable to fetch. Status code : %s" % (str(response.status_code)))
+    elif response.status_code==200:
+        ### If first submission of sample, query will return empty list
+        if len(response.json())==0:
+            return True
+        else:
+        ### Handles instances where sample has existing analyses
+            for analysis in response.json():
+                ### If analysis is suppressed; we ignore
+                if analysis["analysisState"]=="PUBLISHED" and analysis["experiment"]["experimental_strategy"]==metadata.get('experimental_strategy'):
```
Comment (lindaxiang) [inline sanity-check/main.py]: Need to add `analysisType=="sequencing_experiment"`

File: sanity-check/main.py [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+        return_metadata[key]=clinical_metadata[key]
+    return return_metadata
+
+def check_study_exists(metadata,submission_song_url):
+    headers={"accept":"*/*"}
+    endpoint="%s/studies/%s" % (submission_song_url,metadata['program_id'])
+    response=requests.get(endpoint,headers=headers)
+    
+    if response.status_code==404:
+        sys.exit("Program %s does not exist. Please verify program code is correct. Otherwise contact DCC-admin for help to troubleshoot." % (metadata.get('program_id')))
```
Comment (lindaxiang) [inline sanity-check/main.py]: Program %s does not exist in Song.

File: sanity-check/README.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+'api_token' - Valid API token retrievable from Platform 
+'song_url' - Valid Song URL
+'clinical_url' - Valid Clinical URL
+
+## Outputs
+`updated_file_info.tsv`
+#### Successful instance
+
+An `updated_file_info.tsv` is generated
+
```
Comment (lindaxiang) [inline sanity-check/README.md:17]: updated_experiment_info.tsv

**Review by lindaxiang — CHANGES_REQUESTED** [review]

File: sanity-check/README.md (hunk 1 above)
Comment (edsu7) [reply sanity-check/README.md:4]: Updated

File: sanity-check/README.md (hunk 11 above)
Comment (edsu7) [reply sanity-check/README.md:17]: Updated

File: sanity-check/main.py (hunk 10 above)
Comment (edsu7) [reply sanity-check/main.py]: Updated:
https://github.com/icgc-argo/argo-data-submission/blob/2a6ee2ba6bf1716bd58ed3d3e008bc08594d0e9c/sanity-check/main.py#L168

File: sanity-check/main.py (hunk 9 above)
Comment (edsu7) [reply sanity-check/main.py]: Updated:
https://github.com/icgc-argo/argo-data-submission/blob/2a6ee2ba6bf1716bd58ed3d3e008bc08594d0e9c/sanity-check/main.py#L190-L194

File: sanity-check/main.py (hunk 8 above)
Comment (edsu7) [reply sanity-check/main.py]: Updated:
https://github.com/icgc-argo/argo-data-submission/blob/2a6ee2ba6bf1716bd58ed3d3e008bc08594d0e9c/sanity-check/main.py#L66-L70

File: sanity-check/main.py (hunk 7 above)
Comment (edsu7) [reply sanity-check/main.py]: https://github.com/icgc-argo/argo-data-submission/blob/2a6ee2ba6bf1716bd58ed3d3e008bc08594d0e9c/sanity-check/main.py#L44-L45

File: sanity-check/main.py (hunk 6 above)
Comment (edsu7) [reply sanity-check/main.py]: Updated:
https://github.com/icgc-argo/argo-data-submission/blob/2a6ee2ba6bf1716bd58ed3d3e008bc08594d0e9c/sanity-check/main.py#L50

File: sanity-check/main.nf (hunk 5 above)
Comment (edsu7) [reply sanity-check/main.nf]: Updated:
https://github.com/icgc-argo/argo-data-submission/blob/2a6ee2ba6bf1716bd58ed3d3e008bc08594d0e9c/sanity-check/main.nf#L50

File: sanity-check/README.md (hunk 4 above)
Comment (edsu7) [reply sanity-check/README.md]: Updated:
https://github.com/icgc-argo/argo-data-submission/blob/2a6ee2ba6bf1716bd58ed3d3e008bc08594d0e9c/sanity-check/README.md?plain=1#L59

File: sanity-check/README.md (hunk 3 above)
Comment (edsu7) [reply sanity-check/README.md]: Updated:
https://github.com/icgc-argo/argo-data-submission/blob/2a6ee2ba6bf1716bd58ed3d3e008bc08594d0e9c/sanity-check/README.md?plain=1#L1

File: sanity-check/README.md (hunk 2 above)
Comment (edsu7) [reply sanity-check/README.md]: Updated:
https://github.com/icgc-argo/argo-data-submission/blob/2a6ee2ba6bf1716bd58ed3d3e008bc08594d0e9c/sanity-check/README.md?plain=1#L7

**Review by lindaxiang — APPROVED** [review]
LGTM

**Pull request merged**